### ISSUE TITLE
Allow Restart of Containers

### ISSUE THREAD

**Issue by corbs9** [testcontainers/testcontainers-java#606]
Currently, test containers does not allow you to restart a running container without stopping it and removing it completely, followed by a start.

Allowing the functionality to restart a running container which issues a docker restart ensures the testability of a container in a multitude of other ways.

Such an example could relate to custom initialisation scripts which create data, and issuing a restart to the container ensures that this data isn't created twice.

**Comment by rnorth** [comment]
Hi @corbs9 
Thanks for the suggestion - this seems like a good idea. I can imagine there being two possible approaches to implementation inside Testcontainers:

- Just do a stop/start via the docker API
- A more thorough approach that does the above and also uses configured wait strategies etc to make sure the container is ready

I think I'd prefer the latter, but it would be much more work.

Keen to hear others' thoughts here (also contributions would be most welcome 😄 )
Rich

**Comment by corbs9** [comment]
Hi @rnorth, 

I thought initially about just restarting the containers, but it seems that port mapping info, for example, is created once at container creation and wouldn't be updated when a docker restart is issued.

I have forked the project and plan to do the following: 

- In [ResourceReaper](https://github.com/testcontainers/testcontainers-java/blob/c2116ca69b3b01bf19e9177bdde3fda787be24ea/core/src/main/java/org/testcontainers/utility/ResourceReaper.java), split out the stop command into stop and remove
- In [GenericContainer](https://github.com/testcontainers/testcontainers-java/blob/c2116ca69b3b01bf19e9177bdde3fda787be24ea/core/src/main/java/org/testcontainers/containers/GenericContainer.java) move the functionality from `tryStart()` into two methods, `tryCreate()` which will create the container, and `tryStart()` which will start the created container. 
- Once this refactor is done, create a new `restart()` method in GenericContainer which will utilise start but also update any attributes that have changed (for example the ports). 

What are your thoughts on this?

Oli

**Comment by kiview** [comment]
Hi @corbs9,

Sounds like a good idea in general. 
Regarding the `ResourceReaper`: We are currently killing the container, so depending on the use case, a stop would need to be more graceful (thinking about databases here).

The changes in `GenericContainer` also sound good and should be okay, since these methods are `private`.

In addition to the `restart()` method, we could then also add a `pause()` method.  I'm a bit worried about having different semantics then the `docker stop` and `docker pause` commands, but we already kind of have this problem.

**Comment by corbs9** [comment]
HI @kiview,

WRT differing docker semantics, renaming the `start()` command to the more aptly named `run()` would be a breaking change as `GenericContainer` is the main entry point into the application. I'd therefore accept that start would be a synonym for run, and in test containers, start and run perform the same job. I would try and limit the disparity between docker semantics and test containers to this only, and try to match 1-2-1 where possible. 

Good mention on the ResourceReaper. I will perhaps add a method which signifies that this is a graceful operation as opposed to the default kill command. 

WRT the `pause()` method: what is the intention of this? 

Oli

**Comment by bsideup** [comment]
Hi @corbs9,

FYI there is an upcoming huge (from a code respective) change in GenericContainer #600, you might want to wait for it before changing the GenericContainer class :)

**Comment by corbs9** [comment]
Hi @bsideup,

Thanks for the heads up. I made a bit of a start last night, but it was just refactoring `GenericContainer` to separate out start and stop etc. 

Is there an ETA on #600? 

Oli

**Comment by kiview** [comment]
@corbs9 
So assuming we won't change our public API, `stop()` would still stop and remove a container. I of course see the use case for your `restart()` method. A public `pause()` method would however allow to stop the container (meaning `docker stop`), maybe check something on the SUT and then start it again.

Seems nice for some resilience test scenarios, especially for system tests.

**Comment by behrangsa** [comment]
Another use case is simulating an out of service condition in integration service:

_Start containers
Ensure subject under test is working 
Stop one or more containers 
Ensure subject under test gracefully tolerates the fault
Restart the stopped containers 
Ensure subject under test resumes normal operation
Verify state of containers (e.g. database rows) is as expected_

**Comment by kiview** [comment]
Hey @behrangsa, you are right, that's also the use case I was explaining above.

**Comment by behrangsa** [comment]
@kiview docker has a built in [pause command](https://docs.docker.com/engine/reference/commandline/pause/) too. So the pause method added to testcontainers api can call that instead of issuing docker stop.

**Comment by kiview** [comment]
Of course, just look at the discussion in this issue :)

Behrang <notifications@github.com> schrieb am Fr., 30. März 2018, 07:37:

**Comment by chungngoops** [comment]
Hi,
this is my approach to support `restart` container without losing its configuration:
- In `ResourceReaper`: separate `stopContainer(String containerId, String imageName)` into two methods:
  - `shutdown(String containerId, String imageName)` : stop the container.
  - `removeContainer(String containerId)`: removes container with its assosicated volumes.
- In `GenericContainer`: make a `restart()` method that similar to the current `tryStart(Profiler profiler)` but we don't need the beginning part that creates container for the image.

Let me know your thoughts?
ChungNguyen

**Comment by rnorth** [comment]
Not stale - PR pending.

**Comment by chungngoops** [comment]
not stale - I am working on the PR.

**Comment by semistone** [comment]
I use 

  
        public void pause() {
            this.dockerClient.pauseContainerCmd(containerId).exec();
        }

        public void unpause() {
            this.dockerClient.unpauseContainerCmd(containerId).exec();
        }
  
to pause and unpause container

**Comment by asafm** [comment]
What happened to this issue - did you end up adding restart command to GenericContainer? I can't seem to find one.

**Comment by bsideup** [comment]
@asafm since restarting the container [would cause it to re-assign the ports](https://github.com/testcontainers/testcontainers-java/pull/939#issuecomment-490218654) and a few other things, we decided to not have this functionality for now.

What's your use case?

**Comment by asafm** [comment]
Well, I have a small java-agent I wrote couple of years ago - jmx2graphite - which spits out the MBeans to Graphite. It had a known issue I wanted to fix but couldn't find the time: When ever Graphite server got restarted, writes after a while started failing on Broken pipe. So I wanted to simulate it, by restarting the Graphite container:
```
        graphiteContainer.getDockerClient()
                         .restartContainerCmd(graphiteContainer.getContainerId())
                         .exec();
```

Unfortunately, it didn't work, since I couldn't get that broken pipe I wanted.
Stop / start did work, but it changed the ports (restart command changed ports as well), and I had to remain on same port to simulate real life, so I ended binding fixed port outside. 

Long story short, restart won't help my use case :)

**Comment by Shmayro** [comment]
A possible solution to restart a container :
https://stackoverflow.com/a/73489794/6485371
